### PR TITLE
Store old key package reference in MessagesSecretsStore

### LINKED ISSUE #705: Decrypting past epochs with key package references [#717]
The mechanism introduced in #563 to decrypt past epochs doesn't work with the changes from #541.
When #541 is fixed the sender of a message is identified by their key package reference. But this reference can change (e.g. through an update). Application messages from a sender that changed their key package in the meantime can not be decrypted because the leaf can not be identified.

In order to restore the functionality introduced in #563 not only message secrets from past epochs must be kept but the key package references from all members in order to identify the sender of application messages from past epochs.

### PR DESCRIPTION
**Pull request by franziskuskiefer**
The mechanism introduced in #563 to decrypt past epochs requires us to store key package references from the past epochs as well.
Since the sender is identified by its key package reference, which can change, this identifier needs to be stored for past epochs and used in lookups.

Fixes #705

### CODE REVIEW

**Review by kkohbrok — CHANGES_REQUESTED** [review]
Thanks for bringing this validation check back! For my understanding this looks good, with just a few comments.

File: openmls/src/group/core_group/mod.rs [hunk 1]
```diff
@@ -659,6 +659,24 @@ impl CoreGroup {
...
+        if epoch < self.context().epoch() {
+            self.message_secrets_store
+                .secrets_and_leaves_for_epoch_mut(epoch)
+                .ok_or({
+                    CoreGroupError::MlsCiphertextError(MlsCiphertextError::SecretTreeError(
+                        SecretTreeError::TooDistantInThePast,
+                    ))
+                })
+        } else {
+            Ok((self.message_secrets_store.message_secrets_mut(), Vec::new()))
```
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply openmls/src/group/core_group/mod.rs:677]: It doesn't matter what we return here as this is never used. If we're interested in the current epoch, we have the tree.

File: openmls/src/group/core_group/mod.rs [hunk 2]
```diff
@@ -659,6 +659,24 @@ impl CoreGroup {
         }
     }
 
+    /// Get the message secrets. Either from the secrets store or from the group.
+    pub(crate) fn message_secrets_and_leaves_mut<'secret, 'group: 'secret>(
```
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply openmls/src/group/core_group/mod.rs:664]: Right, there's no other place where we can easily do this. Having all the logic for handling past epochs in one place is also nice.

File: openmls/src/group/core_group/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kkohbrok) [reply openmls/src/group/core_group/mod.rs:677]: Fair enough, but then that should at least be mentioned in the docs. Without any further information, I would assume from the name of the function that I get the message secrets and the leaves from the given epoch. It's not really clear that the current epoch is a special case.

**Review by kkohbrok — DISMISSED** [review]
Thanks! Let's get this in!

**Review by raphaelrobert — APPROVED** [review]
Thanks! I had to look at this a bit longer because of the validation logic (see me inline comment).

File: openmls/src/framing/validation.rs [hunk 3]
```diff
@@ -134,17 +151,38 @@ impl DecryptedMessage {
     ///  - ValSem246
     ///  - Prepares ValSem247 by setting the right credential. The remainder
     ///    of ValSem247 is validated as part of ValSem010.
-    pub fn credential(&self, treesync: &TreeSync) -> Result<Credential, ValidationError> {
+    pub fn credential(
```
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply openmls/src/framing/validation.rs:154]: 👍🏻 I filed #754 and put in the validation meta issue to have a note for potential follow ups.

**Review by raphaelrobert — APPROVED** [review]
Thanks!

**Review by kkohbrok — APPROVED** [review]

**Pull request merged**